numpy/core/shape_base.py

```diff
@@ -179,6 +179,86 @@ def atleast_3d(*arys):
         return res
 
 
+def atleast_nd(arr, n, front=False):
+    r"""
+    View inputs as arrays with at least n dimensions.
+
+    Parameters
+    ----------
+    arr : array_like
+        One array-like object.  Non-array inputs are converted to arrays.
+        Arrays that already have n or more dimensions are preserved.
+
+    n : int
+        number of dimensions to ensure
+
+    front : bool
+        if True new dimensions are added to the front of the array.  otherwise
+        they are added to the back.
+
+    Returns
+    -------
+        ndarray :
+            An array with ``a.ndim >= n``.  Copies are avoided where possible,
+            and views with three or more dimensions are returned.  For example,
+            a 1-D array of shape ``(N,)`` becomes a view of shape
+            ``(1, N, 1)``, and a 2-D array of shape ``(M, N)`` becomes a view
+            of shape ``(M, N, 1)``.
+
+    See Also
+    ---------
+    atleast_1d, atleast_2d, atleast_3d
+
+    Examples
+    --------
+    >>> for n in range(0, 5):
+    ...    print(np.atleast_nd(3.0, n))
+    3.0
+    [3.]
+    [[3.]]
+    [[[3.]]]
+    [[[[3.]]]]
+
+    >>> arr = [[1, 2, 3], [4, 5, 6]]
+    >>> for n in range(0, 5):
+    ...    new = np.atleast_nd(arr, n, front=False)
+    ...    print(new.tolist())
+    [[1, 2, 3], [4, 5, 6]]
+    [[1, 2, 3], [4, 5, 6]]
+    [[1, 2, 3], [4, 5, 6]]
+    [[[1], [2], [3]], [[4], [5], [6]]]
+    [[[[1]], [[2]], [[3]]], [[[4]], [[5]], [[6]]]]
+
+    >>> arr = [[1, 2, 3], [4, 5, 6]]
+    >>> for n in range(0, 5):
+    ...    new = np.atleast_nd(arr, n, front=True)
+    ...    print(new.tolist())
+    [[1, 2, 3], [4, 5, 6]]
+    [[1, 2, 3], [4, 5, 6]]
+    [[1, 2, 3], [4, 5, 6]]
+    [[[1, 2, 3], [4, 5, 6]]]
+    [[[[1, 2, 3], [4, 5, 6]]]]
+
+    >>> x = np.arange(12.0).reshape(4,3)
+    >>> np.atleast_nd(x, 5, front=False).shape
+    (4, 3, 1, 1, 1)
+    >>> np.atleast_nd(x, 5, front=True).shape
+    (1, 1, 1, 4, 3)
+    >>> np.atleast_nd(x, 5, front=False).base is x.base  # x is a reshape, so not base itself
+    True
+    """
+    arr_ = asanyarray(arr)
```

Comment:
Why keep the original reference around?

Reply:
This is a stylistic preference. If possible I prefer to never overwrite a reference to an object. This has advantages if you are like me and always develop with an open IPython terminal and a doctest. I don't think there's any strong reason to change it.

Reply:
Not overwriting references leads to a greater memory usage, as all the references have to stick around till the end of the function - if you replace a reference, it can be freed immediately

Reply:
True, but when arr is an ndarray, arr_ will still have a reference to the data, so at best you are freeing a thin wrapper.  I think in the vast majority of circumstances any gains are marginal enough to be ignored, hence why I think its valid to say its a stylistic preference. I defiantly will not claim that the other way is wrong. Its a moot point anyway.